### PR TITLE
Navbar Notifications

### PR DESCRIPTION
**Pull request by AnkushMalik**
## Changes :

- From now on, whenever a user creates a project/joins a project he will be automatically be added as member of project_followers(watcher).
- Notifications will be created whenever a person joins/likes or create a comment on a project.

## Screenshots :
screen shot 2018-07-07 at 4 52 00 pm

screen shot 2018-07-07 at 4 53 36 pm

older notifications page:
image

### CODE REVIEW

File: db/migrate/20180702104632_create_notifications.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class CreateNotifications < ActiveRecord::Migration
```
Comment (differentreality) [inline db/migrate/20180702104632_create_notifications.rb]: Migrating gives an error:
```
Directly inheriting from ActiveRecord::Migration is not supported. Please specify the Rails release the migration was written for:
  class CreateNotifications < ActiveRecord::Migration[4.2]
```
This is due to the Rails 5 update. You can add the text needed by hand, or migrate down this migration, and create a new migration.

File: app/controllers/comments_controller.rb [hunk 2]
```diff
@@ -12,6 +12,7 @@ def create
     @comment.commenter = current_user
 
     if @comment.save
+      @comment.send_notification current_user, " commented on #{@comment.project.aasm_state}: #{@comment.project.title}"
```
Comment (DavidKang) [inline app/controllers/comments_controller.rb]: missing parenthesis

File: app/controllers/projects_controller.rb [hunk 3]
```diff
@@ -106,6 +107,8 @@ def recess
   # PUT /projects/1/join
   def join
     if @project.join!(current_user)
+      @project.send_notification current_user, "just joined #{@project.aasm_state}: #{@project.title}"
```
Comment (DavidKang) [inline app/controllers/projects_controller.rb]: missing parenthesis

File: app/controllers/projects_controller.rb [hunk 4]
```diff
@@ -126,6 +129,7 @@ def leave
   # PUT /projects/1/like
   def like
     @project.like! current_user
+    @project.send_notification current_user, "just liked #{@project.aasm_state}: #{@project.title}"
```
Comment (DavidKang) [inline app/controllers/projects_controller.rb]: missing parenthesis

File: app/models/notification.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+class Notification < ActiveRecord::Base
+  belongs_to :recipient, class_name: 'User'
+  belongs_to :actor, class_name: 'User'
+  belongs_to :notifiable, polymorphic: true
+  
+  scope :unread, ->{where read_at: nil}
```
Comment (DavidKang) [inline app/models/notification.rb]: Please add space between brackets

File: app/models/project.rb [hunk 6]
```diff
@@ -196,6 +196,14 @@ def similar_projects
     return similar = Project.current(@episode).active.select {|project| (project.keywords & keywords).any?}
   end
 
+  def send_notification sender, message
```
Comment (DavidKang) [inline app/models/project.rb]: missing parenthesis

File: app/views/layouts/_notification_tab.html.haml [hunk 7]
```diff
@@ -0,0 +1,29 @@
+%ul.navbar-notifications.nav.navbar-nav.navbar-right.hidden-xs
+  %li.dropdown
+    - if @notifications.any?
+      = link_to mark_as_read_notifications_path(), id: 'notification-dropdown-activator',method: :post, data: { 'toggle': 'dropdown', 'behavior': 'notifications' }, remote: true do
```
Comment (DavidKang) [inline app/views/layouts/_notification_tab.html.haml]: missing space in between `method`

File: app/views/layouts/_notification_tab.html.haml [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+      %h5 Notifications
+      %hr
+      %ul#notification-dropdown
+        -if @notifications.any?
+          - @notifications.each do |notification|
+            %li{ 'data-behavior':'notifications-link' }
+              = link_to project_path(notification.notifiable) do
+                = image_tag notification.notifiable.avatar.url(:thumb)
+                %p
+                  #{ notification.actor.name } #{ notification.action }
```
Comment (DavidKang) [inline app/views/layouts/_notification_tab.html.haml]: Please don't add space inside the insterpolation, this should be `#{notification.actor.name} #{notification.action}`

File: app/views/notifications/index.html.haml [hunk 9]
```diff
@@ -0,0 +1,14 @@
+%h3 Your Notifications
+%hr
+%br
+
+.older-notifications
+  - @older_notifications.each do |notification|
+    = link_to project_path(notification.notifiable) do
+      %p
+        = image_tag notification.notifiable.avatar.url(:thumb), width: 40
+        #{ notification.actor.name } #{ notification.action }
```
Comment (DavidKang) [inline app/views/notifications/index.html.haml]: Please don't add space inside the insterpolation, this should be `#{notification.actor.name} #{notification.action}`

File: app/views/notifications/index.html.haml [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+        %span.pull-right (#{ time_ago_in_words(notification.created_at) } ago)
```
Comment (DavidKang) [inline app/views/notifications/index.html.haml]: Please don't add space inside the insterpolation.

File: app/models/project.rb [hunk 11]
```diff
@@ -196,6 +196,14 @@ def similar_projects
     return similar = Project.current(@episode).active.select {|project| (project.keywords & keywords).any?}
   end
 
+  def send_notification(sender, message)
+    recipients = project_followers - [sender]
+    recipients.each do |recipient|
+      Notification.create(recipient: recipient, actor: sender, action: message, notifiable: self)
+    end
+  end
+
```
Comment (DavidKang) [inline app/models/project.rb]: Please remove this line

File: app/models/notification.rb [hunk 12]
```diff
@@ -0,0 +1,7 @@
+class Notification < ActiveRecord::Base
+  belongs_to :recipient, class_name: 'User'
+  belongs_to :actor, class_name: 'User'
+  belongs_to :notifiable, polymorphic: true
+
+  scope :unread, ->{ where read_at: nil }
```
Comment (DavidKang) [inline app/models/notification.rb]: missing parenthesis in `where`

File: app/controllers/notifications_controller.rb [hunk 13]
```diff
@@ -0,0 +1,12 @@
+class NotificationsController < ApplicationController
+  before_action :authenticate_user!
+
+  def index
+    @older_notifications = Notification.where(recipient: current_user).order('created_at DESC').page(params[:page])
```
Comment (DavidKang) [inline app/controllers/notifications_controller.rb]: `where(recipient: current_user)` should be a scope.

File: app/models/project.rb (hunk 11 above)
Comment (AnkushMalik) [reply app/models/project.rb]: removed

**Pull request merged**